### DISCUSSION TITLE
how can I get the all  the items of the  chunk after the the item skipped, Thanks!

### DISCUSSION THREAD

**Discussion by kenshin614** [spring-projects/spring-batch#4579] [Q&A] [answered]
**Question description**

**Scenario**：1000 items need to be processed, chunkSize is 10, after the first Item  in the chunk meet the skip condition and happened , how can i get all the 10 items including the above skipped one  in the chunk. as we know if any item in the chunk happened skip,the whole chunk will rollback. so i need to get the all the items in the chunk so as to re-process them in the next step. In the skiplistener, I can only get the only skipped item,not the whole chunk items.

 pseudo java code like this,
`<batch:step parent="stepTemplate" id="testSkip" next="stopSkip">
			<batch:partition partitioner="xxPartitioner">
				<batch:step parent="stepTemplate">
					<batch:tasklet>
						<batch:chunk reader="skipDemoReader" processor="skipDemoProcessor" writer="skipDemoWriter" commit-interval="10"
									 skip-limit="3" >
							<batch:skippable-exception-classes>
								<batch:include class="MySkipException"/>
								<batch:exclude class="java.lang.Exception"/>
							</batch:skippable-exception-classes>
							
							<batch:listeners>
								<batch:listener ref="mySkipListener"></batch:listener>
							</batch:listeners>
						</batch:chunk>
					</batch:tasklet>
				</batch:step>
			</batch:partition>
		</batch:step>`

**Environment**
java 8, springbatch 4.1.1 ,mysql5.7

**Comment by kenshin614** [comment]
@fmbenhassine would you please give me some tips？currently,from the source class of FaultTolerantChunkProcessor.java,I can only get the skipped item,not the chunk items.
`private void callSkipListeners(final Chunk<I> inputs, final Chunk<O> outputs) {

		for (SkipWrapper<I> wrapper : inputs.getSkips()) {
			I item = wrapper.getItem();
			if (item == null) {
				continue;
			}
			Throwable e = wrapper.getException();
			callProcessSkipListener(item, e);
		}

		for (SkipWrapper<O> wrapper : outputs.getSkips()) {
			Throwable e = wrapper.getException();
			try {
				getListener().onSkipInWrite(wrapper.getItem(), e);
			}
			catch (RuntimeException ex) {
				throw new SkipListenerFailedException("Fatal exception in SkipListener.", ex, e);
			}
		}

		// Clear skips if we are possibly going to process this chunk again
		outputs.clearSkips();
		inputs.clearSkips();

	}`

**Comment by fmbenhassine** [comment, marked as answer]
It is not possible to get the entire chunk using a `SkipListener`, because this listener is called when the chunk is scanned item by item to determine the faulty one.

Since you need to get all the items in the chunk to re-process them in the next step, you can use an `ItemWriteListener#onWriteError`. This callback gives you access to the entire chunk when an error happens, and can be used to store items in a staging area that the next step can access. Have you considered this option?

**Comment by fmbenhassine** [comment]
I believe my last comment answers the question, so I am closing this discussion. But if you need more help on this, please add a comment.
